pkg/dependenciesdistributor/dependencies_distributor.go

```diff
@@ -0,0 +1,551 @@
+			binding.GetNamespace(), binding.GetName(), err)
+		return err
+	}
+
+	// create or update attached bindings
+	var errs []error
+	for _, dependent := range dependencies {
+		resource := workv1alpha2.ObjectReference{
+			APIVersion: dependent.APIVersion,
+			Kind:       dependent.Kind,
+			Namespace:  dependent.Namespace,
+			Name:       dependent.Name,
+		}
+
+		rawObject, err := helper.FetchWorkload(d.DynamicClient, d.InformerManager, d.RESTMapper, resource)
+		if err != nil {
+			// do nothing if resource template not exist.
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			errs = append(errs, err)
+			continue
+		}
+
+		attachedBinding := buildAttachedBinding(binding, rawObject)
+		if err := d.createOrUpdateAttachedBinding(attachedBinding); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (d *DependenciesDistributor) recordDependenciesForIndependentBinding(binding *workv1alpha2.ResourceBinding, dependencies []configv1alpha1.DependentObjectReference) error {
+	dependenciesBytes, err := json.Marshal(dependencies)
+	if err != nil {
+		klog.Errorf("failed to marshal dependencies of binding(%s/%s): %v", binding.Namespace, binding.Name, err)
+		return err
+	}
+
+	objectAnnotation := binding.GetAnnotations()
+	if objectAnnotation == nil {
+		objectAnnotation = make(map[string]string, 1)
+	}
+
+	objectAnnotation[bindingDependenciesAnnotationKey] = string(dependenciesBytes)
+
+	return retry.RetryOnConflict(retry.DefaultRetry, func() (err error) {
+		binding.SetAnnotations(objectAnnotation)
+		updateErr := d.Client.Update(context.TODO(), binding)
+		if updateErr == nil {
+			return nil
+		}
+
+		updated := &workv1alpha2.ResourceBinding{}
+		if err = d.Client.Get(context.TODO(), client.ObjectKey{Namespace: binding.Namespace, Name: binding.Name}, updated); err != nil {
+			//make a copy, so we don't mutate the shared cache
+			binding = updated.DeepCopy()
+		} else {
+			klog.Errorf("failed to get updated binding %s/%s: %v", binding.Namespace, binding.Name, err)
+		}
+		return updateErr
+	})
+}
+
+func (d *DependenciesDistributor) findOrphanAttachedResourceBindings(independentBinding *workv1alpha2.ResourceBinding, dependencies []configv1alpha1.DependentObjectReference) ([]*workv1alpha2.ResourceBinding, error) {
+	attachedBindings, err := d.listAttachedBindings(independentBinding.Namespace, independentBinding.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	dependenciesSets := sets.NewString()
+	for _, dependent := range dependencies {
+		key := dependent.Kind + dependent.APIVersion + dependent.Namespace + dependent.Name
+		dependenciesSets.Insert(key)
+	}
+
+	var orphanAttachedBindings []*workv1alpha2.ResourceBinding
+	for _, attachedBinding := range attachedBindings {
+		key := attachedBinding.Spec.Resource.Kind + attachedBinding.Spec.Resource.APIVersion + attachedBinding.Spec.Resource.Namespace + attachedBinding.Spec.Resource.Name
```

Comment:
The key production mode can be abstracted into a method to facilitate subsequent modification of the key production mode.